src/Simplex/Chat.hs

```diff
@@ -2858,77 +2886,152 @@ processChatCommand' vr = \case
+          timed_ <- sndContactCITimed live ct itemTTL
+          -- all quotedItems_ are Nothing in case of multi send - see assertMultiSendable
+          (msgContainers, quotedItems_) <- L.unzip <$> prepareMsgs (L.zip cmrs fInvs_) timed_
+          msgs_ <- sendDirectContactMessages user ct $ L.map XMsgNew msgContainers
+          let itemsData = prepareSndItemsData msgs_ cmrs ciFiles_ quotedItems_
+          (errs, cis) <- partitionEithers <$> saveSndChatItems user (CDDirectSnd ct) itemsData timed_ live
+          unless (null errs) $ toView $ CRChatErrors (Just user) errs
+          forM_ (timed_ >>= timedDeleteAt') $ \deleteAt ->
+            forM_ cis $ \ci ->
+              startProximateTimedItemThread user (ChatRef CTDirect contactId, chatItemId' ci) deleteAt
+          pure $ CRNewChatItems user (map (AChatItem SCTDirect SMDSnd (DirectChat ct)) cis)
           where
-            quoteData :: ChatItem c d -> CM (MsgContent, CIQDirection 'CTDirect, Bool)
-            quoteData ChatItem {meta = CIMeta {itemDeleted = Just _}} = throwChatError CEInvalidQuote
-            quoteData ChatItem {content = CISndMsgContent qmc} = pure (qmc, CIQDirectSnd, True)
-            quoteData ChatItem {content = CIRcvMsgContent qmc} = pure (qmc, CIQDirectRcv, False)
-            quoteData _ = throwChatError CEInvalidQuote
-    sendGroupContentMessage :: User -> GroupId -> Bool -> Maybe Int -> ComposedMessage -> Maybe CIForwardedFrom -> CM ChatResponse
-    sendGroupContentMessage user groupId live itemTTL (ComposedMessage file_ quotedItemId_ mc) itemForwarded = do
+            setupSndFileTransfers :: CM (NonEmpty (Maybe FileInvitation, Maybe (CIFile 'MDSnd)))
+            setupSndFileTransfers =
+              forM cmrs $ \(ComposedMessage {fileSource = file_}, _) -> case file_ of
+                Just file -> do
+                  fileSize <- checkSndFile file
+                  (fInv, ciFile) <- xftpSndFileTransfer user file fileSize 1 $ CGContact ct
+                  pure (Just fInv, Just ciFile)
+                Nothing -> pure (Nothing, Nothing)
+            prepareMsgs :: NonEmpty (ComposeMessageReq, Maybe FileInvitation) -> Maybe CITimed -> CM (NonEmpty (MsgContainer, Maybe (CIQuote 'CTDirect)))
+            prepareMsgs cmsFileInvs timed_ =
+              forM cmsFileInvs $ \((ComposedMessage {quotedItemId, msgContent = mc}, itemForwarded), fInv_) ->
+                case (quotedItemId, itemForwarded) of
+                  (Nothing, Nothing) -> pure (MCSimple (ExtMsgContent mc fInv_ (ttl' <$> timed_) (justTrue live)), Nothing)
+                  (Nothing, Just _) -> pure (MCForward (ExtMsgContent mc fInv_ (ttl' <$> timed_) (justTrue live)), Nothing)
+                  (Just qiId, Nothing) -> do
+                    CChatItem _ qci@ChatItem {meta = CIMeta {itemTs, itemSharedMsgId}, formattedText, file} <-
+                      withFastStore $ \db -> getDirectChatItem db user contactId qiId
```

Comment:
it doesn't seem critical to allow quote + multi select because UI would be complicated as well, so api prohibits it and doesn't bother to batch here. all other message preparation steps except this are pure - so we only do single db read in case of 1 quoted item